sdk/dotnet/Pulumi/Serialization/ResourcePackages.cs

```diff
@@ -1,40 +1,98 @@
-// Copyright 2016-2019, Pulumi Corporation
+// Copyright 2016-2020, Pulumi Corporation
 
 using System;
-using System.Collections.Concurrent;
 using System.Collections.Generic;
+using System.Collections.Immutable;
 using System.Diagnostics.CodeAnalysis;
-using Google.Protobuf.Collections;
-using Google.Protobuf.WellKnownTypes;
+using System.Linq;
+using System.Reflection;
+using Semver;
 
 namespace Pulumi
 {
     public interface IResourcePackage
     {
-        ProviderResource ConstructProvider(string name, string type, IDictionary<string, object?>? args, string urn);
+        string Name { get; }
+        string? Version { get; }
+        ProviderResource ConstructProvider(string name, string type, string urn);
+        Resource Construct(string name, string type, string urn);
```

Comment:
Just a note that coupling these operations--provider and resource construction--constrains Pulumi SDKs s.t. they are delivered in the same assembly. I _think_ that's fine, but I want to call it out explicitly.

Reply:
One alternative that we might want to consider is marking resource classes with an attribute that records their type token and then mapping from type token to class. I think that approach is additive, and we could switch to it in the future without any breaks.

Reply:
These are brilliant points. I went ahead and drafted an alternative implementation based on attributes. I fired a separate PR https://github.com/pulumi/pulumi/pull/5875 to be able to discuss both. I find it simpler, more flexible, and more idiomatic. PTAL cc @justinvp

Reply:
I also like the idea of using an attribute to record the type token. I'll take a look at the other PR.